platform/ext/target/nordic_nrf/common/core/tfm_hal_its_encryption.c

```diff
@@ -0,0 +1,288 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "platform/include/tfm_platform_system.h"
+#include "platform/include/tfm_hal_its.h"
+#include "psa/crypto.h"
+
+#include "nrf_cc3xx_platform_derived_key.h"
+#include "nrf_cc3xx_platform.h"
+#include "nrf_cc3xx_platform_kmu.h"
+#include <string.h>
+
+#define ITS_ENCRYPTION_SUCCESS 0
+
+#define HUK_KMU_SLOT 2
+#define HUK_KMU_SIZE_BITS 128
+
+/* Global encryption counter which resets per boot. The counter ensures that
+ * the nonce will not be identical for consecutive file writes during the same
+ * boot.
+ */
+static uint32_t g_enc_counter;
+
+/* The global nonce seed which is fetched once in every boot. The seed is used
+ * as part of the nonce and allows the platforms to diversify their nonces
+ * across resets. Note that the way that this seed is generated is platform
+ * specific, so the diversification is optional.
+ */
+static uint8_t g_enc_nonce_seed[TFM_ITS_ENC_NONCE_LENGTH -
+                                sizeof(g_enc_counter)];
+
+
+enum tfm_hal_status_t tfm_hal_its_aead_generate_nonce(uint8_t *nonce,
+                                                      size_t nonce_size)
+{
+    int err;
+
+    if(nonce == NULL){
+        return TFM_HAL_ERROR_INVALID_INPUT;
+    }
+
+    if(nonce_size < sizeof(g_enc_nonce_seed) + sizeof(g_enc_counter)){
+        return TFM_HAL_ERROR_INVALID_INPUT;
+    }
+
+    if (g_enc_counter == 0) {
+        if (sizeof(g_enc_nonce_seed) != 8) {
+            return TFM_HAL_ERROR_GENERIC;
+        }
+
+        err = nrf_cc3xx_platform_get_nonce_seed(g_enc_nonce_seed);
+        if (err != 0) {
+            return TFM_HAL_ERROR_GENERIC;
+        }
+    }
+
+    memcpy(nonce, g_enc_nonce_seed, sizeof(g_enc_nonce_seed));
+    memcpy(nonce + sizeof(g_enc_nonce_seed),
+               &g_enc_counter,
+               sizeof(g_enc_counter));
+
+    g_enc_counter++;
+
+    return TFM_HAL_SUCCESS;
+}
```

Comment:
```suggestion
psa_status_t status = psa_generate_random(nonce, nonce_size);
return psa_status_to_tfm_hal_status(status);
```

Do we not have access to PSA APIs?

Reply:
I gues @Vge0rge will have a better answer on it than me so leaving this open until he stated his design decisions

Reply:
We don't have access to the PSA APIs, no :/ 

The reason for this is that the PSA crypto service stores the persistent keys in the ITS partition. So the ITS partition cannot go back to the PSA crypto service since this creates a cyclic dependency which is not accepted.

So we are forced to use platform specific APIs for this.

But even if we just implement a platform specific functionality which does an equivalent thing I think that it needs to be justified why we need to do the extra effort here and add the security advantages for this.

Reply:
Ok then i would leave it @SebastianBoe you can resolve this conversation if you are fine with it

Reply:
This must be documented for future reference.

Reply:
TF-M doesn't allow cyclic dependencies.

Reply:
Or to be more specific the PSA Firmware Framework doesn't allow circular dependencies. You can search for "circular" in the framework.

Reply:
I added a short comment